ykllvmwrap/src/jitmodbuilder.cc

```diff
@@ -1109,6 +1109,25 @@ createModuleForTraceCompilerTests(Module *AOTMod, char *FuncNames[],
   JITModBuilder JB = JITModBuilder::CreateMocked(
       AOTMod, FuncNames, BBs, TraceLen, FAddrKeys, FAddrVals, FAddrLen);
   auto JITMod = JB.createModule();
+
+  // When the trace compiler encounters a non-const global in a trace, it
+  // inserts an LLVM `global external` variable referencing the variable in the
+  // interpreter's address space. In the trace compiler tests, such global
```

Comment:
Is this only a problem in tests? Should this be feature gated if so?

Reply:
It's only an issue for tests, and the function that I've changed is [already gated](https://github.com/ykjit/yk/pull/493/files/18beb6706c3e0e1ba56be8fa3d0025c7110f7d58#diff-b76ed8a29ed650e7c2a44a5809bc05a22cca958d0b9f5532dab74cce8503d112L1103), so we should be grand.

Reply:
The link didn't work as expected, but if you expand the diff upwards, you will see the function is guarded with `#ifdef YK_TESTING`.

Reply:
Well, it's something that can arise in real interpreters, and I wanted to write tests that demonstrate that they are correctly externed (and not copied), otherwise writes to them might not be observed by non-jit-compiled code.

I think we should allow non-const globals in tests.

Reply:
> [I'm probably missing something obvious!]

It's not obvious -- hurts my head too!

In a real interpreter you might have a C program like:
```
int my_global_int = 4;

int main() {
  // interpreter loop
  ...
  my_global_int++;
}
```

The global becomes an llvm global variable in the AOT module.

If we collect a trace that writes to that global, then when we execute that trace, mutations need to be to be observed by non-jit-compiled-code, should we ever exit the trace. We insert an extern variable reference to `my_global_int` in the JIT module and when we compile the trace, the extern matches a symbol in the process initiating the JIT compilation (i.e. the interpreter), and all is dandy.

[If the global is constant, we just inline it -- I'll be testing that at some point too]

Here's the crux of it though: in trace compiler tests, we provide a hypothetical AOT module, which may talk about global variables and whatnot, **but those variables don't really exist**. The hypothetical AOT module hasn't actually ever been code-genned and the process initiating the JIT compilation **is the test harness**. Obviously the test harness doesn't define symbols found in the hypothetical AOT module, and stuff explodes.

To demonstrate, the tests in this PR use a global variable `g` in the hypothetical AOT module. When we compile a trace which manipulates `g`, LLVM looks for a symbol called `g` in the current address space (that of the test harness) and doesn't find one, and borks.

This change looks for these "otherwise problematic" globals and tells LLVM (only when testing):

> mate, trust me: that variable is at address `0xdeadbeef`

This is obviously semantically incorrect, should we execute the trace, but for these kinds of tests, we only check that they compile.

Does it make sense?

Reply:
OK, I get it (I think).

> This is obviously semantically incorrect, should we execute the trace, but for these kinds of tests, we only check that they compile.

OK, so I guess my concern is that `jitmodbuilder.cc` isn't just about "tests that we compile but don't run" -- doesn't this PR mean that if at some point in the future we use it for tests that could run, things will go wrong in weird ways that we'll struggle to track down?

Reply:
The way we fake the control point already means that we can never run the traces generated by this kind of test.

If we want to run traces, we should either use a C program test, or we can make the kind of test suite we discussed where you paste a whole post-our-passes LLVM AOT module into a file and make traces from that.

Reply:
I'm still nervous that we have this weird hack in that's only for testing but it's for *all* testing. If, say, the hack only kicked in if you specifiy `YK_TESTING_GLOBAL_HACK` (or whatever) it would make the hacks existence a bit clearer, and scoped in nature?

Reply:
I hear ya, but this code path is only executed for the trace compiler test suite. See, the hack is inside `createModuleForTraceCompilerTests()`.

Reply:
OK, fair point.